core/src/memory/binary_page_memory_resource_impl.cpp

```diff
@@ -14,24 +14,35 @@
 // System include(s).
 #include <algorithm>
 #include <memory>
-#include <stack>
+#include <optional>
 #include <stdexcept>
 
+#ifdef _MSC_VER
+#include <intrin.h>
+#endif
+
 namespace {
 /**
- * @brief Rounds a size up to the nearest power of two.
+ * @brief Rounds a size up to the nearest power of two, and returns the power
+ * (not the size itself).
  */
 std::size_t round_up(std::size_t size) {
-    for (unsigned short i = 0; i <= 32; i++) {
-        std::size_t s = 2UL << i;
-
-        if (s >= size) {
-            return s;
+    for (std::size_t i = 0; i < 32; i++) {
+        if ((static_cast<std::size_t>(1UL) << i) >= size) {
+            return i;
         }
     }
 
     return 0;
 }
+
+std::size_t __inline clzl(std::size_t i) {
+#ifdef _MSC_VER
+    return _lzcnt_u64(i);
+#else
+    return __builtin_clzl(i);
+#endif
+}
```

Comment:
As discussed in person, let me give an alternative implementation here.

I would much prefer for CMake to make the decision on which function the code should use internally. With something like:

```cmake
include( CheckCXXSymbolExists )
check_cxx_symbol_exists( "__lzcnt_u64" "intrin.h" VECMEM_HAVE_LZCNT_U64 )
if( VECMEM_HAVE_LZCNT_U64 )
   target_compile_definitions( vecmem_core
      PRIVATE VECMEM_HAVE_LZCNT_U64 )
endif()
check_cxx_symbol_exists( "__builtin_clzl" "" VECMEM_HAVE_BUILTIN_CLZL )
if( VECMEM_HAVE_BUILTIN_CLZL )
   target_compile_definitions( vecmem_core
      PRIVATE VECMEM_HAVE_BUILTIN_CLZL )
endif()
```

And then the code could look like:

```c++
std::size_t clzl(std::size_t i) {
#if defined(VECMEM_HAVE_LZCNT_U64)
    return _lzcnt_u64(i);
#elif defined(VECMEM_HAVE_BUILTIN_CLZL)
    return __builtin_clzl(i);
#else
    // Hand-written implementation for the function...
#endif
}
```

Note that I have some reservations against `__inline` as well. :thinking: Can't we just use `inline` instead?

https://en.cppreference.com/w/cpp/language/inline

If not, then we'll need to figure out with CMake also whether `__inline` is available from the compiler being used...